euclidlib/le3/pk_wl.py

```diff
@@ -281,6 +222,98 @@ def mixing_matrices(path: str | PathLike[str]) -> dict[_DictKey, Result]:
+
+    if os.path.exists(path):
+        os.remove(path)
+
+    with fitsio.FITS(path, "rw") as fits:
+        for key, result in results.items():
+            name = _key_to_string(key)
+            arr = np.asarray(result.array)
+            axis = result.axis
+            if isinstance(axis, int):
+                axis = (axis,)
+            elif axis is None:
+                axis = ()
+            if arr.ndim == 1:
+                nrows = arr.shape[0]
+                reshaped_arr = arr
+                tdim = None
+            elif arr.ndim in (2, 3):
+                dim_shape = arr.shape[:-1]
+                nrows = arr.shape[-1]
+                reshaped_arr = arr.reshape(-1, nrows).T
+                tdim = f"({','.join(map(str, dim_shape[::-1]))})"
+            else:
+                raise ValueError(f"Unsupported array shape: {arr.shape}")
+
+            def get_tuple_or_default(
+                attr: str, default_dtype: np.dtype[Any]
+            ) -> np.ndarray[Any, Any]:
+                val = getattr(result, attr, None)
+                if val is None:
+                    return np.zeros(nrows, dtype=default_dtype)
+                val = np.asarray(val)
+                if isinstance(val, (tuple, list)):
+                    val = np.stack([np.asarray(v) for v in val], axis=-1)
+                return val.reshape(nrows).astype(default_dtype)
+
+            ell = get_tuple_or_default("ell", np.dtype(np.int64))
+            lower = get_tuple_or_default("lower", np.dtype(np.int64))
+            upper = get_tuple_or_default("upper", np.dtype(np.int64))
+            weight = get_tuple_or_default("weight", np.dtype(np.float64))
+
+            array_shape = (
+                tuple(map(int, tdim.strip("()").split(",")))
+                if tdim
+                else reshaped_arr.shape[1:]
+            )
+
+            dtype = [
+                ("ARRAY", "f8", array_shape),
+                ("ELL", "i8"),
+                ("LOWER", "i8"),
+                ("UPPER", "i8"),
+                ("WEIGHT", "f8"),
+            ]
+
+            structured_array = np.empty(nrows, dtype=dtype)
+            structured_array["ARRAY"] = reshaped_arr.reshape((nrows,) + array_shape)
+            structured_array["ELL"] = ell
+            structured_array["LOWER"] = lower
+            structured_array["UPPER"] = upper
+            structured_array["WEIGHT"] = weight
+
+            header = {
+                "ELLAXIS": repr(axis),
+            }
+            if tdim is not None:
+                header["TDIM1"] = tdim
+            if meta := arr.dtype.metadata:
+                for k, v in meta.items():
+                    header[f"META {k.upper()}"] = str(v)
+
+            # Write the header to the FITS file
+            history = getattr(_, "_history", None)
+            if history is not None:
+                header["HISTORY"] = history
+            fits.write(structured_array, extname=name, header=header)
```

Comment:
We want to make sure that this function does the same thing this [one](https://github.com/heracles-ec/heracles/blob/2639cf15816b53f948dbf6d40205897d56a0f313/heracles/io.py#L510-L537) and this [one](https://github.com/heracles-ec/heracles/blob/2639cf15816b53f948dbf6d40205897d56a0f313/heracles/io.py#L229-L293) currently inside heracles. These functions divide the writing process into two steps, a function `write` that loops through the keys of the dictionary, and `_write_result` that saves each result to fits. If you collapse both into one function it should give you what you want here. It would also be great if all the other functions follow a similar logic but with different field names.